File: nodes/message-role-swap/message-role-swap.process.js
export default async ({inputs, settings, config, nodeConfig}) => {
  try {
    const messages = inputs.messages;

    if (!Array.isArray(messages)) {
      throw new Error("Messages input must be an array");
    }

    const swappedMessages = [];

    for (const message of messages) {
      if (!message || typeof message !== 'object') {
        // Keep non-object messages as-is
        swappedMessages.push(message);
        continue;
      }

      // Create a copy of the message to avoid mutating the original
      const swappedMessage = { ...message };

      // Swap roles: assistant ↔ user
      if (message.role === 'assistant') {
        swappedMessage.role = 'user';
      } else if (message.role === 'user') {
        swappedMessage.role = 'assistant';
      }
      // Leave other roles (system, tool, developer) unchanged

      swappedMessages.push(swappedMessage);
    }

    return {
      messages: swappedMessages
    };

  } catch (error) {
    throw new Error(`Message Role Swap error: ${error.message}`);
  }
};

File: nodes/message-role-swap/message-role-swap.process.py
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        messages = inputs.get('messages')

        if not isinstance(messages, list):
            raise Exception("Messages input must be an array")

        swapped_messages = []

        for message in messages:
            if not message or not isinstance(message, dict):
                # Keep non-object messages as-is
                swapped_messages.append(message)
                continue

            # Create a copy of the message to avoid mutating the original
            swapped_message = message.copy()

            # Swap roles: assistant ↔ user
            if message.get('role') == 'assistant':
                swapped_message['role'] = 'user'
            elif message.get('role') == 'user':
                swapped_message['role'] = 'assistant'
            # Leave other roles (system, tool, developer) unchanged

            swapped_messages.append(swapped_message)

        return {
            'messages': swapped_messages
        }

    except Exception as error:
        raise Exception(f"Message Role Swap error: {str(error)}")
